### PR TITLE
[KYUUBI #449] Add Basic Metrics Service framework

### PR DESCRIPTION
**Pull request by yaooqinn**
yaooqinn [Closes #450](https://github.com/yaooqinn/kyuubi/pull/450)    [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=yaooqinn&utm_campaign=badge_info)

### _Why are the changes needed?_

This PR adds a future-proofing Metrics Service framework close #499

add some reporters, like json file, jmx, etc..
add some basic metrics, like statement.error

### _How was this patch tested?_
- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

WIP
- [ ] Add screenshots for manual tests if appropriate

- [ ] [Run test](https://kyuubi.readthedocs.io/en/latest/tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

**Comment by yaooqinn** [comment]
Cc @zwangsheng @pan3793 @ulysses-you, can you check this proposal?

File: externals/kyuubi-spark-sql-engine/pom.xml [hunk 1]
```diff
@@ -133,6 +133,18 @@
...
+
+        <dependency>
+            <groupId>com.fasterxml.jackson.core</groupId>
+            <artifactId>jackson-databind</artifactId>
+            <scope>test</scope>
+        </dependency>
+
+        <dependency>
+            <groupId>com.fasterxml.jackson.module</groupId>
+            <artifactId>jackson-module-scala_2.12</artifactId>
```
Comment (pan3793) [inline externals/kyuubi-spark-sql-engine/pom.xml:145]: replace 2.12 by `${scala_binary_version}`

Comment (yaooqinn) [reply externals/kyuubi-spark-sql-engine/pom.xml:145]: nice catch~

**Comment by pan3793** [comment]
Can we also support prometheus?

**Comment by yaooqinn** [comment]
> Can we also support prometheus?

the current way works for prometheus, e.g. using jmx. 

Do you have any better idea?

File: kyuubi-metrics/src/main/scala/org/apache/kyuubi/metrics/MetricsSystem.scala [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+import org.apache.kyuubi.config.KyuubiConf
+import org.apache.kyuubi.metrics.MetricsConf.METRICS_REPORTERS
+import org.apache.kyuubi.metrics.MetricsSystem.maybeSystem
+import org.apache.kyuubi.metrics.ReporterType._
+import org.apache.kyuubi.service.CompositeService
+
+class MetricsSystem extends CompositeService("MetricsSystem") {
+
+  private val registry = new MetricRegistry
+  private var counters: LoadingCache[String, Counter] = _
```
Comment (pan3793) [inline kyuubi-metrics/src/main/scala/org/apache/kyuubi/metrics/MetricsSystem.scala:36]: Is the `LoadingCache` necessary? What if just use `registry.counter(key)`?

**Comment by pan3793** [comment]
> > Can we also support prometheus?
> 
> the current way works for prometheus, e.g. using jmx.
> 
> Do you have any better idea?

I mean a `PrometheusReporter` which export dropwizard metrics to a HTTP restful api and return prometheus syntax data. But it seems not officially supported by dropwizard metrics framework.

**Comment by yaooqinn** [comment]
Hi @pan3793 I am going the merge this, and I am also happy to see your PR with `micrometer`

**Pull request closed without merging**